### ISSUE TITLE
Allow for alternative inherited YAML reaction definition

### ISSUE THREAD

**Issue by ischoegl** [Cantera/cantera#1028]
**Problem description** / **Steps to reproduce**

The indenting of the `reaction` field for 'inherited' reactions is somewhat non-intuitive. While this works:

```yaml
phases:
- name: gas
  species:
  - h2o2.yaml/species: [H, O2, H2O, HO2]
  thermo: ideal-gas
  kinetics: gas
  reactions:
  - h2o2.yaml/reactions: declared-species
```

the more intuitive version (with reactions indented as in other YAML files) would be

```yaml
phases:
- name: gas
  species:
  - h2o2.yaml/species: [H, O2, H2O, HO2]
  thermo: ideal-gas
  kinetics: gas
reactions:
- h2o2.yaml/reactions: declared-species
```

which, however, throws an error

```
***********************************************************************
InputFileError thrown by AnyMap::operator[]:
Error on line XY of input string:
Key 'equation' not found.
Existing keys: h2o2.yaml/reactions
|  Line |
|     4 |       species:
|     5 |       - h2o2.yaml/species: [H, O2, H2O, HO2]
|     6 |       thermo: ideal-gas
|     7 |       kinetics: gas
|     8 |     reactions:
>     9 >     - h2o2.yaml/reactions: declared-species
                ^
|    10 |     
***********************************************************************
```

The error does point out the issue, but as there never was an intent to provide an equation, it is not necessarily intuitive.

**System information**

- Cantera version: 2.6.0a2

**Comment by bryanwweber** [comment]
I think this is intentional. There is a distinction between the `reactions` field of the `phases` mapping and the top-level `reactions` key, which is the name of a section which will be used by default to find reaction specifications if no other field is specified. It is possible to combine both reactions from another file and reactions in the `reactions` key in the same phase specification. This is similar to the `species` field.

**Comment by ischoegl** [comment]
> I think this is intentional. 

Yes, I do realize that there is a distinction as different things are accomplished. But allowing for the alternative isn't too difficult to implement, and would be imho more intuitive. It's certainly a papercut ...

**Comment by decaluwe** [comment]
But the `reactions` at question here is a sub-field of the `phase` declaration.  I.e. “these are the reactions that correspond to the phase currently being declared.” I don’t see the purpose of un-indenting this field. As @bryanwweber puts it, it is a parallel construct to the other member fields (.e.g. `species`, `thermo`, etc.) not sure how’d I’d ever justify having it at a different indentation level as those parallel fields.

**Comment by bryanwweber** [comment]
@decaluwe I can see a case where one might want to declare this at the top level, if multiple phases were to include the same set of reactions imported from another file. So I definitely see a possibility there. That would be a change from existing CTI behavior, which is fine, but I think why I didn't consider that case before.

I can't recall whether `reactions: all` is required in the phase definition when using the default top-level `reactions` field. If it is, I don't think this kind of construct would save any cognitive overhead, so I'd only be in favor of improving the error message, if possible.

**Comment by decaluwe** [comment]
Hmm, I think I see that.  Fwiw, I don’t think one needs to declare `reactions: all`. I could be wrong - I’ll follow up on this, later (on a mobile, device, right now).

**Comment by speth** [comment]
While the visible difference in this example is only the indentation of this block, there's a clear distinction between the two that I think is worth preserving. The `reactions` field of a phase entry specifies where reactions are defined, and a top-level `reactions` field provides reaction definitions.

Introducing the option to add a layer of indirection here doesn't seem to me like something that would actually see much use -- you'd need to be defining multiple phases in the input file that all wanted to use the same set of reactions, and where that set of reaction definitions was complex enough that you didn't want to duplicate it in each phase definition.

Furthermore, using the top-level `reactions` field in this way would introduce some unfortunate parsing ambiguity, because we would now have to distinguish between the two cases based on the actual content rather than just the structure of the data. Specifically, I think the most direct way to tell that:
```yaml
reactions:
- h2o2.yaml/reactions: declared-species
```
isn't a reaction definition is because it doesn't include an `equation` field, which doesn't exactly lend itself to robust or clear error handling.

**Comment by bryanwweber** [comment]
@ischoegl @speth @decaluwe Any further thoughts here? It seems like there's limited support for this change, so shall we close the issue?

**Comment by ischoegl** [comment]
Before we close it, I wanted to briefly summarize some thoughts I had on this.

The way I understand the code base, there is a difference between `species` and `reaction` definitions, so there is imho not a clear argument to treat them as parallel constructs (it may be so from the old CTI perspective, but I don't have that institutional memory).

For gas-phase reactions, it is clear that reactions are associated with a certain phase, although in an *indirect* fashion. This is less clear for reactions at surfaces etc., where a reaction involves species associated with multiple phases. This is also clear from a coding perspective, where reactions are associated with `Kinetics` objects that link to a *vector* of `ThermoPhase` objects, see

https://github.com/Cantera/cantera/blob/01bf417683a015fe85393273c9f5aab8ebc32a0a/include/cantera/kinetics/Kinetics.h#L895-L908

In other words, reactions are primarily associated with species, and only indirectly with phases. From that perspective, I believe that it would make sense to allow for the unindented import I suggested above.

**PS:** At the same time, it is still true that there is a need to define a *primary* phase for a `Kinetics` object, with others being *adjacent* (see `Solution` in Python) ... so some of my comments above are moot.
